edb/server/server.py

```diff
@@ -788,22 +808,71 @@ def _on_remote_system_config_change(self):
     def _on_global_schema_change(self):
         self._loop.create_task(self._reintrospect_global_schema())
 
-    def _on_sys_pgcon_connection_lost(self):
+    def _on_sys_pgcon_connection_lost(self, exc):
         if not self._serving:
-            # The server is shutting down.
+            # The server is shutting down, release all events so that
+            # the waiters if any could continue and exit
+            self._sys_pgcon_ready_evt.set()
+            self._sys_pgcon_reconnect_evt.set()
             return
+
+        logger.error(
+            "Connection to the system database is " +
+            ("closed." if exc is None else f"broken! Reason: {exc}")
+        )
+        self.set_pg_unavailable_msg(
+            "Connection is lost, please check server log for the reason."
+        )
         self.__sys_pgcon = None
+        self._sys_pgcon_ready_evt.clear()
+        self._loop.create_task(self._reconnect_sys_pgcon())
 
-        async def reconnect():
-            conn = await self._pg_connect(defines.EDGEDB_SYSTEM_DB)
-            if self.__sys_pgcon is not None or not self._serving:
-                conn.abort()
+    async def _reconnect_sys_pgcon(self):
+        try:
+            conn = None
+            while self._serving:
+                try:
+                    conn = await self._pg_connect(defines.EDGEDB_SYSTEM_DB)
+                    break
+                except ConnectionError:
+                    # Keep retrying as far as:
+                    #   1. The EdgeDB server is still serving,
+                    #   2. We still cannot connect to the Postgres cluster, or
+                    pass
+                except pgcon_errors.BackendError as e:
+                    #   3. The Postgres cluster is still starting up
+                    if e.fields['C'] != '57P03':
```

Comment:
Hm, looks like we're now checking these codes in multiple places. How about we add methods to the BackedErorr type, like `is_something()`?

Reply:
Discussed offline, conclusion included in the next CRF